tac/agents/participant/agent.py

```diff
@@ -103,17 +104,19 @@ def react(self) -> None:
         counter = 0
         while (not self.mailbox.inbox.empty() and counter < self.max_reactions):
             counter += 1
-            msg = self.mailbox.inbox.get_nowait()  # type: Optional[Envelope]
-            logger.debug("processing message of protocol={}".format(msg.protocol_id))
-            if msg is not None:
-                if is_oef_message(msg):
-                    self.oef_handler.handle_oef_message(msg)
-                elif is_controller_message(msg, self.crypto):
-                    self.controller_handler.handle_controller_message(msg)
-                elif is_fipa_message(msg):
-                    self.dialogue_handler.handle_dialogue_message(msg)
+            envelope = self.mailbox.inbox.get_nowait()  # type: Optional[Envelope]
+            logger.debug("processing message of protocol={}".format(envelope.protocol_id))
+
+            if envelope is not None:
+                if is_oef_message(envelope):
+                    self.oef_handler.handle_oef_message(envelope)
+                elif is_controller_message(envelope):
+                    self.controller_handler.handle_controller_message(envelope)
+                elif is_fipa_message(envelope):
+                    fipa_message = FIPASerializer().decode(envelope.message)  # type: # Message
```

Comment:
I'd move the decode into the handler; so the interface is the same for all three